### ISSUE TITLE
PlatformException(AudioWaveforms, Failed to prepare player, null, null)

### ISSUE THREAD

**Issue by appinteractive** [SimformSolutionsPvtLtd/audio_waveforms#264]
Hey, when I try to load an audio file, I get the following exception in Xcode:

```
         MP3AudioFile.cpp:1526   Problem scanning for packets
         MP3AudioFile.cpp:1065   MPEGAudioFile::OpenFromDataSource failed
      AudioFileObject.cpp:105    OpenFromDataSource failed
       AudioFileObject.cpp:80    Open failed
          ExtAudioFile.cpp:211   about to throw 'dta?': open audio file
          AVAEInternal.h:109   [AVAudioFile.mm:135:AVAudioFileImpl: (ExtAudioFileOpenURL((CFURLRef)fileURL, &_extAudioFile)): error 1685348671
           ```
           
In Flutter itself I get `PlatformException(AudioWaveforms, Failed to prepare player, null, null)`

The issue happens when I try to extract the waveform, but setting it in the `initializePlacer` to false does not help.

I tried several mp3`s, does someone have one that works?

**Comment by priyankgandhi0** [comment]
I also facing this issue @DevarshRanpara can help me out ?

**Comment by appinteractive** [comment]
Funny enough, on Android it does work. Only iOS (15) is exploding.

**Comment by DevarshRanpara** [comment]
Hello @appinteractive 
Can you please share your implementation code and plugin version (from pubspec.lock file).
We will look into this ASAP.

**Comment by priyankgandhi0** [comment]
@DevarshRanpara this is my pubspec.lock file 

  audio_waveforms:
    dependency: "direct main"
    description:
      name: audio_waveforms
      sha256: f2f81e256c5c03b64e2861e0323a77e1b6790c256c73393b671214c9846260fc
      url: "https://pub.dev"
    source: hosted
    version: "1.0.4"
    
    ==> Code
      void _initialiseRecordControllers() {
    recorderController = RecorderController()
      ..androidEncoder = AndroidEncoder.aac
      ..androidOutputFormat = AndroidOutputFormat.ogg
      ..iosEncoder = IosEncoder.kAudioFormatMPEG4AAC
      ..sampleRate = 44100;
    maxDuration = const Duration(seconds: 10);
    _initialisePlayerController();
  }
  
  For record =>  await recorderController.record();
  For stop record =>  audioPath = await recorderController.stop();

**Comment by appinteractive** [comment]
It works now

My findings so far are:

- I had no luck with the docs, could not get it running
- using paths from the assets folder does not work, you have to write the files to an app dir (was not obvious to me)
- I had to copy the whole pub.dev example and get the chat_bubble from the examples folder to get it working
- now I need to pull out the parts that I need

Thanks for your quick response @DevarshRanpara. I think the docs are just quite incomplete. Also the file in the example folder does import `import 'package:audio_waveforms_example/chat_bubble.dart';` which does not exist when you try to get it working in the first place as I was trying to import it.

**Comment by yokawaiik** [comment]
The same problem. I've used example code for voice player widget with  wave.

How can we fix this problem.

**Comment by appinteractive** [comment]
> The same problem. I've used example code for voice player widget with  wave.
> 
> 
> 
> How can we fix this problem. 

How do you load the audio file? It needs to be on the device before being able to generate. 

For me also using files on the assets directory did not work. It seems only files in the app data or temp directory work. 

For files on the assets folder you could pre-compile the values as it's more more performant any way. 

@DevarshRanpara maybe that should be mentioned in the docs!?

**Comment by yokawaiik** [comment]
@appinteractive 

I have the following code:

`

import 'package:flutter/material.dart';
// Begin custom widget code
// DO NOT REMOVE OR MODIFY THE CODE ABOVE!

// Set your widget name, define your parameter, and then add the
// boilerplate code using the green button on the right!

import 'dart:io';
import 'dart:async'
import 'package:flutter_cache_manager/flutter_cache_manager.dart';
import 'package:audio_waveforms/audio_waveforms.dart' as audio_waveforms;
import 'package:http/http.dart' as http;

class VoiceAudioLoaderWithPlayer extends StatefulWidget {
  const VoiceAudioLoaderWithPlayer({
    Key? key,
    this.width,
    this.height,
    required this.sourceUrl,
    required this.errorString,
    this.isCurrentUser = false,
  }) : super(key: key);

  final String sourceUrl;
  final String errorString;
  final bool isCurrentUser;

  final double? width;
  final double? height;

  @override
  State<VoiceAudioLoaderWithPlayer> createState() =>
      _VoiceAudioLoaderWithPlayer();
}

class _VoiceAudioLoaderWithPlayer extends State<VoiceAudioLoaderWithPlayer> {
  String? _audioFilePath;

  Future<File> _loadAudioFileByUrl(String url) async {
    try {
      // final DefaultCacheManager cacheManager = DefaultCacheManager();
      final fetchedFile = await DefaultCacheManager().getSingleFile(url);
      // final fetchedFile = _downloadFile(url, getFileName(url));
      // final fetchedFile = await cacheManager.downloadFile(url);

      debugPrint('fetchedFile: ${fetchedFile}');
      // cacheManager.dispose();

      return fetchedFile;
    } catch (e) {
      debugPrint('_loadAudioFileByUrl - e: $e');

      rethrow;
    }
  }

  Future<void> _loadPlayer() async {
    try {
      if (_audioFilePath != null) {
        return;
      }
      final audioFile = await _loadAudioFileByUrl(widget.sourceUrl);
      _audioFilePath = audioFile.path;
    } catch (e) {
      debugPrint('_loadPlayer - e: $e');
      rethrow;
    }
  }

  @override
  void dispose() {
    debugPrint('VoiceAudioLoaderWithPlayer - dispose: $_audioFilePath');
    super.dispose();
  }

  late final double width;
  late final double height;

  // if didnt do it, component will bebuild multiple times
  late final Future loadPlayerFuture;

  @override
  void initState() {
    width = widget.width ?? 140;
    height = widget.height ?? 40;

    loadPlayerFuture = _loadPlayer();

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: loadPlayerFuture,
      builder: (
        BuildContext context,
        AsyncSnapshot snapshot,
      ) {
        if (snapshot.hasError) {
          return VoiceAudioError(errorString: widget.errorString);
        }

        if (snapshot.connectionState == ConnectionState.done) {
          return SizedBox(
            height: height,
            width: width,
            child: Row(
              children: [
                AudioWaveformPlayer(
                  key: ValueKey(_audioFilePath),
                  path: _audioFilePath!,
                  isCurrentUser: widget.isCurrentUser,
                  waveWidth: width - 58,
                  waveHeight: height! - 22.0,
                ),
              ],
            ),
          );
        } else if (snapshot.connectionState == ConnectionState.waiting) {
          return Row(
            children: [
              Container(
                  height: height,
                  width: width,
                  decoration: BoxDecoration(
                    border: widget.isCurrentUser
                        ? Border.all(
                            color: Colors.grey.shade300,
                            width: 1.0,
                          )
                        : null,
                    borderRadius: BorderRadius.only(
                      topLeft: Radius.zero,
                      topRight: Radius.circular(16),
                      bottomLeft: Radius.circular(16),
                      bottomRight: Radius.circular(16),
                    ),
                  ),
                  child: Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 8.0),
                    child: Row(
                      children: [
                        CustomLoaderSpinner(
                          size: 28,
                          color: Colors.grey,
                        ),
                      ],
                    ),
                  )),
            ],
          );
        } else {
          debugPrint('snapshot.connectionState: ${snapshot.connectionState}');
          return SizedBox.shrink();
        }
      },
    );
  }
}

class VoiceAudioError extends StatelessWidget {
  const VoiceAudioError({
    super.key,
    required this.errorString,
    this.color,
    this.textColor,
  });

  final String errorString;
  final Color? color;
  final Color? textColor;

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.maxFinite,
      height: 50,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(12),
        color: color ?? Colors.grey.shade100,
      ),
      child: Center(
        child: Text(
          errorString,
          style: TextStyle(
              color: textColor ?? Colors.black,
              fontSize: 16,
              fontWeight: FontWeight.w600),
        ),
      ),
    );
  }
}

//
//
// AudioWaveformPlayer
//
//

const _kIconPlay = Icon(
  Icons.play_arrow_rounded,
  size: 26.0,
  color: Colors.white,
);

const _kIconPause = Icon(
  Icons.pause_rounded,
  size: 26.0,
  color: Colors.white,
);

class AudioWaveformPlayer extends StatefulWidget {
  const AudioWaveformPlayer({
    Key? key,
    required this.path,
    this.isCurrentUser = false,
    this.playIcon,
    this.pauseIcon,
    this.fixedWaveColor,
    this.liveWaveColor,
    this.waveWidth = 140,
    this.waveHeight = 34,
  }) : super(key: key);

  final String path;
  final bool isCurrentUser;
  final Icon? playIcon;
  final Icon? pauseIcon;
  final Color? fixedWaveColor;
  final Color? liveWaveColor;
  final double waveWidth;
  final double waveHeight;

  @override
  State<AudioWaveformPlayer> createState() => _AudioWaveformPlayerState();
}

class _AudioWaveformPlayerState extends State<AudioWaveformPlayer> {
  late audio_waveforms.PlayerController _playerController;

  late StreamSubscription<audio_waveforms.PlayerState> _playerStateSubscription;

  // bool loading = false;
  Duration? _duration;
  late Icon _icon;

  @override
  void initState() {
    _icon = widget.playIcon ?? _kIconPlay;
    _duration = Duration.zero;

    super.initState();
    _playerController = audio_waveforms.PlayerController();

    _preparePlayer();
    _playerStateSubscription =
        _playerController.onPlayerStateChanged.listen((_) {
      setState(() {});
    });
  }

  statusChange(audio_waveforms.PlayerController playerController) async {
    if (playerController.playerState ==
        audio_waveforms.PlayerState.initialized) {
      playerController.startPlayer(
          forceRefresh: true, finishMode: audio_waveforms.FinishMode.pause);
    
      _icon = widget.pauseIcon ?? _kIconPause;

    }

    if (playerController.playerState == audio_waveforms.PlayerState.playing) {
      playerController.pausePlayer();
    
      _icon = widget.playIcon ?? _kIconPlay;

    }

    if (playerController.playerState == audio_waveforms.PlayerState.paused) {
      playerController.startPlayer(
          forceRefresh: true, finishMode: audio_waveforms.FinishMode.pause);
  
      _icon = widget.pauseIcon ?? _kIconPause;
   
    }

    if (playerController.playerState == audio_waveforms.PlayerState.stopped) {
      playerController.startPlayer(
          forceRefresh: true, finishMode: audio_waveforms.FinishMode.pause);
    
      _icon = widget.pauseIcon ?? _kIconPause;
  
    }

    // final playerStateSubscription =
    await playerController.onPlayerStateChanged.listen((event) async {
      if (event == audio_waveforms.PlayerState.paused) {
    
        _icon = widget.playIcon ?? _kIconPlay;
   
      }
    });
  }

  void _preparePlayer() async {
    final samples = const audio_waveforms.PlayerWaveStyle()
        .getSamplesForWidth(widget.waveWidth);

   _playerController.preparePlayer(
      path: widget.path,
      noOfSamples: samples,
      shouldExtractWaveform: true,
    );
    
    // todo: get duration 
    // final durationInMiliSeconds = await _playerController.getDuration();
    // final _duration = Duration(milliseconds: durationInMiliSeconds);



    setState(() {
      
    });,
  }

  // _initPlayer() async {
  //   playerController = audio_waveforms.PlayerController();

  String _formatDuration(Duration duration) {
    String minutes =
        duration.inMinutes.remainder(60).toString().padLeft(2, '0');
    String seconds =
        duration.inSeconds.remainder(60).toString().padLeft(2, '0');
    return '$minutes:$seconds';
  }

  Future<Duration?> _getMaxAudioDuration(String filePath) async {
    try {

      final durationInMiliSeconds = await _playerController.getDuration();
      final duration = Duration(milliseconds: durationInMiliSeconds);
      return duration;
    } catch (e) {
      debugPrint('_getMaxAudioDuration - e: $e');
      return null;
    }
  }


  @override
  void dispose() {
    _playerStateSubscription.cancel();
    _playerController.stopPlayer();
    _playerController.dispose();

    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(horizontal: 6, vertical: 2),
      decoration: BoxDecoration(
        border: widget.isCurrentUser
            ? Border.all(
                color: widget.fixedWaveColor ?? Colors.grey.shade300,
                width: 1.0,
              )
            : null,
        borderRadius: BorderRadius.only(
          topLeft: Radius.zero,
          topRight: Radius.circular(16),
          bottomLeft: Radius.circular(16),
          bottomRight: Radius.circular(16),
        ),
      ),
      child: Column(
        children: [
          Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              Material(
                type: MaterialType.transparency,
                child: Ink(
                  decoration: BoxDecoration(
                    border: Border.all(color: Colors.white, width: 2.0),
                    shape: BoxShape.circle,
                  ),
                  child: InkWell(
                    //This keeps the splash effect within the circle
                    borderRadius: BorderRadius.circular(
                        1000.0), //Something large to ensure a circle
                    onTap: () {
                      statusChange(_playerController);
                    },
                    child: Padding(
                      padding: EdgeInsets.all(2),
                      child: _icon,
                    ),
                  ),
                ),
              ),
              SizedBox(width: 6),
              Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  audio_waveforms.AudioFileWaveforms(
                    key: ValueKey(widget.path),
                    size: Size(
                      widget.waveWidth,
                      widget.waveHeight,
                    ),
                    playerController: _playerController,
                    enableSeekGesture: true,
                    waveformType: audio_waveforms.WaveformType.fitWidth,
                    waveformData: _playerController.waveformData,
                    playerWaveStyle: audio_waveforms.PlayerWaveStyle(
                      fixedWaveColor:
                          widget.fixedWaveColor ?? Colors.grey.shade500,
                      liveWaveColor: widget.liveWaveColor ?? Colors.white,
                      spacing: 6,
                      showTop: true,
                      scaleFactor: 100,
                      scrollScale: 3,
                      seekLineThickness: 4,
                      showBottom: true,
                      seekLineColor: Colors.black87,
                      showSeekLine: true,
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 4, bottom: 4),
                    child: Text(
                      // _formatDuration(_duration ?? Duration.zero),
                      // todo: если проигрывается, то показывать длительность проигрывания
                      // todo:
                      // todo: если не проигрывается, то показывать длительность записи

                      _formatDuration(_duration ?? Duration.zero),
                      style: TextStyle(
                        fontSize: 10,
                        color: Colors.white,
                        letterSpacing: 1,
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                    // todo: показывать длительность записи в стриме
                        StreamBuilder<int>(
                      stream: _playerController.onCurrentDurationChanged,
                      builder: (context, snapshot) {
                        final textStyle = TextStyle(
                          fontSize: 10,
                          color: Colors.white,
                          letterSpacing: 1,
                          fontWeight: FontWeight.w500,
                        );

                        if (snapshot.hasData) {
                          return Text(
                            _formatDuration(
                              // Duration(milliseconds: snapshot.data!)
                              //  ??
                              Duration.zero,
                            ),
                            style: textStyle,
                          );
                        } else {
                          return Text(
                            _formatDuration(
                              Duration(milliseconds: 0),
                            ),
                            style: textStyle,
                          );
                        }
                      },
                    ),
                  ),
                ],
              ),
            ],
          ),
        ],
      ),
    );
  }
}
`

It's contain two widget:
1. VoiceAudioLoaderWithPlayer - wrapper which loads audiofrom backend
2. AudioWaveformPlayer - player with wave

Current AudioWaveformPlayer has error, but i don't understand what am i wrong.

**Comment by appinteractive** [comment]
@yokawaiik I use the extract method and setting the `shouldExtractWaveform`.


```dart

**Comment by yokawaiik** [comment]
The exception is caused by the `extractWaveformData` method.

<details><summary>Exception in terminal output</summary>
E/flutter ( 2224): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: PlatformException(AudioWaveforms, Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference, An error is thrown before decoding the audio file, null)
  E/flutter ( 2224): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:651:7)
  E/flutter ( 2224): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:354:18)
  E/flutter ( 2224): <asynchronous suspension>
  E/flutter ( 2224): #2      AudioWaveformsInterface.extractWaveformData (package:audio_waveforms/src/base/audio_waveforms_interface.dart:163:9)
  E/flutter ( 2224): <asynchronous suspension>
  E/flutter ( 2224): #3      PlayerController.extractWaveformData (package:audio_waveforms/src/controllers/player_controller.dart:169:20)
  E/flutter ( 2224): <asynchronous suspension>
  E/flutter ( 2224): #4      _AudioWaveformPlayerState.init.<anonymous closure>.<anonymous closure> (package:out_and_about/custom_code/widgets/voice_audio_loader_with_player.dart:356:68)
  E/flutter ( 2224): <asynchronous suspension>
</details> 

The error pops up in this method which calls a function from the framework
```dart
  Future<List<double>> extractWaveformData({
    required String key,
    required String path,
    required int noOfSamples,
  }) async {
    final result =
        await _methodChannel.invokeMethod(Constants.extractWaveformData, {
      Constants.playerKey: key,
      Constants.path: path,
      Constants.noOfSamples: noOfSamples,
    });
    return List<double>.from(result ?? []);
  }
```

**Comment by yokawaiik** [comment]
Looks like I've solved my problem.

The error was caused by the fact that the files that were loaded onto the device and played through this library with extract waves were loaded incorrectly.
I used Firebasse Storage as storage and, when loading recorded audio, saved it with the specified Metadata, which I had not done before. Apparently, the file was saved to the server incorrectly this way or had incorrect meta tags. Anyway, just removed the code that added Metadata on load.
And in the end, the file was loaded normally with the code that was there and the waves were successfully extracted.

Thus my problem was solved.

**Comment by abul-kalam-mari** [comment]
@yokawaiik I agree with you. I have been trying to solve this issue since 2 weeks.

**Cause of Problem**
When you record and upload audio to server or Firebase its not supported properly. 

Here is what I did:

- I uploaded a song to my server and loaded that audio.mp3 in cached directory on my IOS device. 
- Used the path of cached audio file to prepare player controller

It worked. It throws exception to unsupported types or corrupted audio files.
